File: clients/typescript/sndv.ts
export interface SndvConfig {
    baseUrl: string;
    token: string;
}

export interface Metrics {
    WriteOps: number;
    ReadOps: number;
    WALSyncs: number;
    ActiveMemTable: number;
    // Add other metrics as needed
}

export class SndvClient {
    private baseUrl: string;
    private headers: HeadersInit;

    constructor(config: SndvConfig) {
        this.baseUrl = config.baseUrl.replace(/\/$/, "");
        this.headers = {
            "Content-Type": "application/json",
            "Authorization": config.token
        };
    }

    private async request<T>(method: string, endpoint: string, body?: any): Promise<T> {
        const url = `${this.baseUrl}${endpoint}`;
        const options: RequestInit = {
            method,
            headers: this.headers,
            body: body ? JSON.stringify(body) : undefined
        };

        const res = await fetch(url, options);

        if (!res.ok) {
            if (res.status === 404) {
                throw new Error("Key not found");
            }
            throw new Error(`SNDV Error [${res.status}]: ${res.statusText}`);
        }

        const text = await res.text();
        return JSON.parse(text);
    }

    /**
     * Write a value to the store
     * @param key Unique key
     * @param value String value
     * @param ttl Seconds to live (0 = forever)
     */
    async put(key: string, value: string, ttl: number = 0): Promise<void> {
        await this.request("POST", "/put", { key, value, ttl });
    }

    /**
     * Get a value by key
     * @returns The value string or null if missing
     */
    async get(key: string): Promise<string | null> {
        try {
            const res = await this.request<{ val: string }>("GET", `/get?key=${key}`);
            return res.val;
        } catch (e: any) {
            if (e.message === "Key not found") return null;
            throw e;
        }
    }

    /**
     * Delete a key (Create Tombstone)
     */
    async delete(key: string): Promise<void> {
        await this.request("POST", `/delete?key=${key}`);
    }

    /**
     * Get Server Metrics
     */
    async metrics(): Promise<Metrics> {
        return this.request<Metrics>("GET", "/metrics");
    }
}

// --- Usage Example ---
/*
async function main() {
    const client = new SndvClient({
        baseUrl: "http://localhost:8080",
        token: "YOUR_TOKEN_HERE"
    });

    await client.put("ts_key", "Typescript Rules");
    const val = await client.get("ts_key");
    console.log("Value:", val);
}
*/

File: clients/python/sndv.py
import urllib.request
import urllib.error
import json
import time

class SndvError(Exception):
    def __init__(self, code, message):
        self.code = code
        self.message = message
        super().__init__(f"[HTTP {code}] {message}")

class SndvClient:
    def __init__(self, base_url="http://localhost:8080", token=""):
        self.base_url = base_url.rstrip("/")
        self.token = token
        self.headers = {
            "Content-Type": "application/json",
            "Authorization": token
        }

    def _request(self, method, endpoint, payload=None):
        url = f"{self.base_url}{endpoint}"
        data = None
        if payload is not None:
            data = json.dumps(payload).encode('utf-8')

        req = urllib.request.Request(url, data=data, headers=self.headers, method=method)

        try:
            with urllib.request.urlopen(req) as response:
                if response.status == 204:
                    return None
                body = response.read()
                if not body:
                    return None
                return json.loads(body)
        except urllib.error.HTTPError as e:
            raise SndvError(e.code, e.reason)
        except urllib.error.URLError as e:
            raise SndvError(0, f"Connection Failed: {e.reason}")

    def put(self, key, value, ttl=0):
        """
        Write a value to the store.
        :param key: String key
        :param value: String value
        :param ttl: Time to live in seconds (0 = infinite)
        """
        payload = {"key": key, "value": value, "ttl": ttl}
        self._request("POST", "/put", payload)

    def get(self, key):
        """
        Retrieve a value. Returns None if not found or token invalid.
        """
        try:
            resp = self._request("GET", f"/get?key={key}")
            return resp.get("val") if resp else None
        except SndvError as e:
            if e.code == 404:
                return None
            raise e

    def delete(self, key):
        """
        Delete a key (Tombstone).
        """
        self._request("POST", f"/delete?key={key}")

    def metrics(self):
        """
        Fetch server metrics.
        """
        return self._request("GET", "/metrics")

# --- Example Usage ---
if __name__ == "__main__":
    # Example assumes server is running locally
    # You must paste your actual admin token here to run this directly
    TOKEN = "PASTE_YOUR_TOKEN_HERE" 
    
    if TOKEN == "PASTE_YOUR_TOKEN_HERE":
        print("Please edit the file and set TOKEN to run the example.")
    else:
        client = SndvClient("http://localhost:8080", TOKEN)
        
        print("Writing...")
        client.put("demo_key", "Hello World", 60)
        
        print("Reading...")
        val = client.get("demo_key")
        print(f"Got: {val}")
        
        print("Metrics:", client.metrics())